xicam/SAXS/processing/correction.py

```diff
@@ -1,56 +1,118 @@
-        dark[2] being the gain 1 (least sensitive) dark image.
-    flat : array_like, optional
-        Input array for the flatfield correction. This should be of shape
-        (y, x)
-    gain : tuple, optional
-        These are the gain multiplication factors for the three different
-        gain settings
-    Returns
-    -------
-    array_like
-        Array of corrected images of shape (N, y, x)
-"""
-    # Images are expected to be of type np.uint16 (csxtools.utils.py:94)
-    bitmasked_images = Input(description="", type=np.ndarray, visible=False)
-    # Darks are expected to be of type np.float32 (csxtools.fastccd.images.py:43)
-    dark_images = Input(description="", type=np.ndarray, visible=False)
-    # Flats are expected to be of type np.float32 (csxtools.fastccd.images.py:47)
-    flat_field = Input(description="", type=np.ndarray, default=None, visible=False)
-    gain = Input(description="", type=tuple, visible=False)
-
-    corrected_images = Output(description="", type=np.ndarray)
+
+from xicam.core import msg
+from xicam.plugins import ProcessingPlugin, Input, Output
+
+
+class CorrectImage(ProcessingPlugin):
+    images = Input(name="image", type=np.ndarray)
+    flats = Input(name="flats", type=np.ndarray)
+    darks = Input(name="darks", type=np.ndarray)
+    gains = Input(name="gains", type=tuple, default=(1, 4, 8))
+    corrected_images = Output(name="corrected_images", type=np.ndarray)
 
     def evaluate(self):
-        data = np.asarray(self.bitmasked_images.value.astype(np.uint16))
-        reduced_dark_image = self.dark_images.value
-        darks = None
-        if reduced_dark_image is not None:
-            reduced_dark_image = stackmean(np.asarray(self.dark_images.value, dtype=np.float32))
-            darks = np.array([reduced_dark_image,
-                              np.zeros(shape=reduced_dark_image.shape),
-                              np.zeros(shape=reduced_dark_image.shape)],
-                             dtype=np.float32)
-        self.corrected_images.value = correct_images(data,
-                                                     dark=darks)
-                                                     # flat=self.flat_field.value.astype(np.float32))
+        def correct(array, flats, bkg, gain_map=(1, 4, 8)):
+            # 16-bit unsigned
+            # image: bits 0 - 12
+            # bad:   bit 13
+            # gain:  bits 14-15
+            # gain1: 0b11 (3)
+            # gain2: 0b10 (2)
+            # gain8: 0b00 (0)
+            intensity = np.bitwise_and(0x1FFF, array)
+            bad_flag = np.bitwise_and(0x1, np.right_shift(array, 13))
+            gain = np.bitwise_and(0x3, np.right_shift(array, 14))
+            # map the gain bit values to the gain map indices to get the actual gain values
+            # e.g. 3 -> index 2; 2 -> index 1; 1 -> index 0
+            gain = np.vectorize(partial(lambda a, b: a[int((b + 1) / 2)], gain_map))(gain)
```

Comment:
If gain_map was a mapping dict, you could write it as `np.vectorize(gain_map.get)(gain)`. Looks cleaner?

Reply:
within `evaluate`, i zipped this into a dictionary. i didn't expose it as a dictionary, since the gain byte masks should be static for this detector, right?